Miniupnpd and bittorrent improvements

bobvanderlinden opened the pull request:
###### Motivation for this change

The bittorrent tests was sometimes failing. It seemed to be related to upnp. There was no test yet that checked whether miniupnpd worked correctly. Lastly the miniupnpd service did not setup iptables correctly.

This PR fixes the miniupnpd iptables setup by using the related scripts from Miniupnpd.

It adds a test which checks whether the Miniupnpd service allows port mapping in a `client <> router <> client` structure. 

It fixes and simplifies the bittorrent test by using the miniupnpd service, the transmission service and by reusing the network structure used in thhe miniupnpd test.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

bobvanderlinden commented:
@jgillich You seem to have more experience with miniupnpd than I do. What I'm running into is the following:

```
client1: must succeed: upnpc -l
router# [   18.270599] miniupnpd[578]: chain MINIUPNPD not found
```

`miniupnpd` does not seem to be able to find the iptables chain `MINIUPNPD` as defined here:

https://github.com/NixOS/nixpkgs/blob/0f1de2ea9fd7f92e745a1ace6f24e73ae750d14f/nixos/modules/services/networking/miniupnpd.nix#L61-L70

Do you have any idea what could be the problem here?

jgillich commented:
Not sure, but I've also never gotten it to work right and found it impossible to debug, so I just gave up. It probably should have been marked as broken.

bobvanderlinden commented:
This is related to getting all tests passing for 18.09: https://github.com/NixOS/nixpkgs/issues/45960

bobvanderlinden commented:
@miniupnp Can you shed some light on what miniupnpcd does? And maybe good ways to debug problems? I'm trying to write a test using a few vms to make sure the core functionality of miniupnpcd always works in NixOS.

The following succeeds:
```
upnpc -a ${...} 9000 9000 TCP
```

But then I don't see the any output from:

```
upnpc -l
```

I don't see much happening in miniupnpcd and I don't see iptables changing using:

```
iptables --list-rules
```

Do you have some suggestions on ways to debug this problem?

miniupnp commented:
what are the output log of _miniupnpd_ while you run the` upnpc -a [...] `command ?
you can run _miniupnpd_ with -d option to get more detailed output

bobvanderlinden commented:
@miniupnp Thank you for pushing me in the right direction. I wasn't aware of the `-d` option and found out the problem (amongst others) was that it couldn't determine its external IP address (which is likely because it's inside of a sandbox environment). I set the external address in the configuration and now the upnp test is working correctly!

Next up is the bittorrent test, where I was initially working on ;).

bobvanderlinden commented:
@jgillich the miniupnpd service should now be working correctly. It now uses the shell scripts that Miniupnpd supplies.

@miniupnp during the tests it occurred to me that `upnpc -a ...` exits with exit code 0 when creating a port mapping has failed. Is this intentional? Should I create an issue for this?

@samueldr If/when this gets merged, this probably also needs backporting to 18.09.

bobvanderlinden committed:
miniupnpd: wrap iptables scripts to use correct PATH

bobvanderlinden committed:
nixos: miniupnpd: use iptables scripts

bobvanderlinden committed:
tests: upnp: init test for upnp using miniupnpd / miniupnpc

bobvanderlinden committed:
tests: bittorrent: improve stability

This attempts to improve stability of the test by using existing
services for miniupnpd and transmission.

It also uses explicit addresses for the network interfaces so that the
external IP addresses are valid internet addresses (thus fixing
validation problems from upnpc).

Also disable eth0 from being used to transfer torrents over without that
being the intention.

bobvanderlinden committed:
tests: bittorrent: add bobvanderlinden as maintainer

miniupnp commented:
@bobvanderlinden how did he fail ? Have you the console output ?

jgillich commented on nixos/modules/services/networking/miniupnpd.nix:
```diff
@@ -59,30 +59,12 @@ in
   config = mkIf cfg.enable {
     # from miniupnpd/netfilter/iptables_init.sh
```
You can probably remove this comment (and the other one below).

Great work btw! :+1:

bobvanderlinden commented:
@miniupnp I'm terribly sorry, but I can't reproduce the errors anymore (reflog is gone and I recreated the commits to adhere to nixpkgs standards :().

The biggest confusion was the exit code always being 0. Initially I didn't realize the test framework did not output any stdout/err of the upnpc comand when the exit code is 0, because succeeding commands do not ouput anything in this framework. So it was not showing up output by default. Based on your comment I added `-d` and realized there still was no output, so I made some changes in the test whatever the exit code is. Errors like not adding required parameters also resulted in exit code 0.

While figuring the above out I ended up using `-d` as the first argument. When I figured out how to get output I got a `NoSuchEntryInArray` error. It took me a while to realize `-d` is also a command when using it as the first argument (oops).

I also struggled a bit with the `ext_ip` and the interfaces having internal addresses by default. By chance I picked `240.x.x.x` for the internet addresses, which apparently was also a reserved range (I was not aware of this). Looking at the code of miniupnpd I found a non-reserved range (80.x.x.x).

@jgillich I removed the comments in the new commit set :+1: thanks for noticing.

bobvanderlinden commented:
@samueldr Could you trigger `nixos.tests.upnp` and `nixos.tests.bittorrent` on @grahamc?

samueldr commented:
@GrahamcOfBorg  nixos.tests.upnp nixos.tests.bittorrent

GrahamcOfBorg commented:
<!--REQUEST_ID=7166b6e1-768c-49dc-9b7c-fd086b45d9c4-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.46443&attempt_id=1834d034-c179-47e5-b876-0aab24d5a557)

Attempted: tests.upnp, tests.bittorrent

<details><summary>Partial log (click to expand)</summary><p>

```
killing client2 (pid 598)
killing router (pid 610)
killing tracker (pid 622)
killing client1 (pid 634)
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: Could not remove ctl dir '/build/vde2.ctl': Directory not empty
vde_switch: Could not remove ctl dir '/build/vde1.ctl': Directory not empty
/nix/store/4x0dnz9r0a5mpadga0q0zkaa0811n70l-vm-test-run-upnp
/nix/store/l7bil5j2f8vkcp9jq8zldcpjprim9yy9-vm-test-run-bittorrent
```
</p></details>

bobvanderlinden commented:
:tada: I think this one is ready to merge :+1:
